src/Model/ValueObject/Web/Url.php

```diff
@@ -20,10 +20,31 @@ public function __construct($value)
         $this->guardString($value);
         $this->guardRegexPattern('/\\Ahttp[s]?:\\/\\//', $value);
 
+        $value = $this->encode($value);
+
         if (filter_var($value, FILTER_VALIDATE_URL) === false) {
             throw new \InvalidArgumentException('Given string is not a valid url.');
         }
 
         $this->setValue($value);
     }
+
+    private function encode(string $value): string
+    {
+        // Take into account already encoded urls by first decoding them.
+        $value = urldecode($value);
+
+        // Encode the url but revert various semantic characters.
+        $value = urlencode($value);
+
+        // Revert meaningful characters.
+        $entities = ['%21', '%2A', '%27', '%28', '%29', '%3B', '%3A', '%40', '%26', '%3D', '%2B', '%24', '%2C', '%2F', '%3F', '%25', '%23', '%5B', '%5D'];
+        $replacements = ['!', '*', '\'', '(', ')', ';', ':', '@', '&', '=', '+', '$', ',', '/', '?', '%', '#', '[', ']'];
```

Comment:
This feels like a workaround that might miss some edge cases. Maybe we could use https://www.php.net/manual/en/function.parse-url.php instead to divide the URL in meaningful parts and only decode-encode those that need to be encoded? Afterwards you can convert it into a string again with https://php.uz/manual/en/function.http-build-url.php
Using these functions and only encoding for example the paths and query parameter values feels safer I think?

Or alternatively, can you point to a resource that documents this approach as standardized?

Reply:
Which is based on https://datatracker.ietf.org/doc/html/rfc3986#section-2.2

Reply:
Perfect, if it's based on a standard it should be safe! Maybe just add a comment to point to these links for future reference.